### PR TITLE
Support metric history plotting with Step/Wall/Relative X-axis and multiple metrics

### PR DESCRIPTION
**Pull request by Zangr**
## What changes are proposed in this pull request?

- Support metric history plotting with Step/Wall/Relative X-axis
- Support metric history plotting with multiple metrics
- Preserve selected plotting metrics in url parameter
- Support grouped bar chart for plotting metrics with a single history, switch to line chart when metric with more than 1 history entry is selected.
- Support toggle line smoothness
- Support logarithmic y-axis
- Support line chart point marker toggle
- Refactor out MetricsPlotPanel for more complex future visualization reuse.

Test TBD.

**Compare Metrics**
final-line

**Compare Metrics With One/Many Metric Histories**
final-bar

## How is this patch tested?
 
(TBD)Test will be added. 

### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
- Support metric history plotting with Step/Wall/Relative X-axis
- Support metric history plotting with multiple metrics
- Preserve selected plotting metrics in url parameter
- Support grouped bar chart for plotting metrics with a single history, switch to line chart when metric with more than 1 history entry is selected.
- Support toggle line smoothness
- Support logarithmic y-axis
- Support line chart point marker toggle
- Refactor out MetricsPlotPanel for more complex future visualization reuse.
 
### What component(s) does this PR affect?
 
- [x] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by aarondav** [comment]
Awesome!

File: mlflow/server/js/src/components/MetricsPlotPanel.js
Comment (smurching) [inline mlflow/server/js/src/components/MetricsPlotPanel.js]: I know we're not super consistent about this in MLflow, but would you mind adding a docstring comment for each of these props? e.g. see [this file](https://github.com/mlflow/mlflow/blob/2a2b2bba3fa2f4c8acfd460e4f3cd7f305ba9c33/mlflow/server/js/src/components/ExperimentView.js#L71) for example

File: mlflow/server/js/src/components/MetricsPlotPanel.js
Comment (smurching) [inline mlflow/server/js/src/components/MetricsPlotPanel.js]: Nanonit: `showPoint`? to match what we use in the UI :P

File: mlflow/server/js/src/components/MetricsPlotPanel.js
Comment (smurching) [inline mlflow/server/js/src/components/MetricsPlotPanel.js]: Could `qs.parse(search)['plot_metric_keys']` ever be undefined? I guess not for now, but maybe we should be defensive here? Mainly because `JSON.parse(undefined)` will throw

File: mlflow/server/js/src/components/MetricsPlotView.js
Comment (smurching) [inline mlflow/server/js/src/components/MetricsPlotView.js]: BTW, I'm guessing you probably already discussed this with others - do we intend for this behavior? It seems to show each metric relative to its run start during run comparison, which I guess actually makes sense, but I was confused about what exactly was happening in the default view (relative timestamp as X axis) initially. Maybe we could/should add a tooltip explaining what relative time means? Could defer it to a follow-up PR too.

Also, would it make sense to have a unit for the relative timestamp x axis (so that it's clear it's in ms)?

**Review by smurching — COMMENTED** [review]
Looks great (would be good to add tests as you mentioned), just leaving some small comments

**Comment by dbczumar** [comment]
@Zangr This is awesome!!! One high-level comment: When we click on a single run info page and then select a metric, the resulting metric graph page displays `Default > Run {run_id} > {metric_name}`, but we enable the user to plot more metrics than `{metric_name}`. Do you think it would make more sense to make the header read `Default > Run {run_id} > Metrics` (without naming a metric explicitly).

File: mlflow/server/js/src/components/MetricsPlotView.js
Comment (smurching) [inline mlflow/server/js/src/components/MetricsPlotView.js]: Nit: I think you need a colon or space here, otherwise the metric name gets concatenated with the run key

File: mlflow/server/js/src/components/MetricView.js
Comment (smurching) [inline mlflow/server/js/src/components/MetricView.js]: If we conditionally include this title when there's a single runUuid, we avoid the issue @dbczumar found where it's confusing that the metric comparison page shows a single metric key as its title even when you're comparing multiple metrics.

**Review by smurching — COMMENTED** [review]
LGTM in terms of functionality besides one nit. I also agree with @dbczumar's comment that it'd be better to show "Metrics" instead of a single metric name from the run comparison page.

 Besides those, once we add tests we should be good to merge

File: mlflow/server/js/src/components/MetricView.js
Comment (dbczumar) [reply mlflow/server/js/src/components/MetricView.js]: Even if there's a single runUuid, the metrics view can still be used to visualize multiple metrics for a single run. Therefore, I'm still in favor of "Metrics"

File: mlflow/server/js/src/components/MetricsPlotView.js
Comment (Zangr) [reply mlflow/server/js/src/components/MetricsPlotView.js]: Yes. will add in the test commit.

File: mlflow/server/js/src/components/MetricView.js
Comment (Zangr) [reply mlflow/server/js/src/components/MetricView.js]: Let's conditionally make it either `Metrics` or `{metric_name}` then.

**Review by dbczumar — APPROVED** [review]
LGTM after we rename the metrics history view header to "Metrics", rather than "{metric_key}"

**Comment by Zangr** [comment]
Fixed "Metrics" /  "{metric_key}". Merging.

**Comment by avinish1** [comment]
I am just trying to figure out how can I plot metrics over time(line plot) as shown above.

I am logging **time as a param and accuracy as a metric** for multiple runs, so I have one (time,accuracy) tuple for every run ID.

**Note:-I have only one (time, accuracy) tuple for every runID in an experiment.**

I am **not able to switch to a line plot view of accuracy over time**

I might be off about this request to help with the same.

**Comment by jsanjay63** [comment]
Hi, Thank you very much for adding this feature. I would like to understand if there's any specific requirement to be able to visualize the metrics in a line plot as seen above. I could visualize only the bar plot as attached. Appreciate your help.
Screenshot 2021-06-29 at 7 56 15 PM

**Pull request merged**
